File: rust/checks/string_states.py
"""Rust magic string checks — from rust/types.md.

Strings used as discriminators (node types, operation kinds, tab IDs)
should be enums. Raw string comparisons and match arms scatter state
definitions across the codebase.

BANNED:
  - match expr { "foo" => ... }  — match on string literal (use enum)
  - if x == "foo"  where the string looks like an identifier/kind value

WARNING:
  - String literals ending in .json/.toml/.png used outside a const/static
    definition — centralise in a paths module (see rust/hardcoded_paths.py)

Exempt:
  - Test files and #[cfg(test)] blocks
  - Error messages and log strings (contain spaces or sentence structure)
  - Single-char strings
  - Lines that ARE const/static definitions
"""

from __future__ import annotations

from pathlib import Path
from typing import Generator
import re

from common.issue import Issue, Severity

_RULE_BASE = "rust/types"

# match arm on a string literal:  "foo" =>
_MATCH_STR_ARM = re.compile(r'"([^"]{2,}?)"\s*=>')

# == "identifier-like"  or  != "identifier-like"
_EQ_STR = re.compile(r'[!=]=\s*"([^"]{2,}?)"')

# const / static definition line — these ARE the source of truth
_CONST_DEF = re.compile(r"^\s*(?:pub\s+)?(?:const|static)\s+")

# Strings that are obviously messages/paths/sentences (not identifiers)
_SKIP_VALUES = re.compile(
    r"(?:\s|[/\\.]|[A-Z]{2}|[?!,])"  # spaces, path chars, SCREAMING, punctuation
)


def _is_test_file(path: Path) -> bool:
    parts = path.parts
    return (
        "tests" in parts
        or "test" in parts
        or path.stem.endswith("_test")
        or path.stem.startswith("test_")
    )


def _is_test_context(lines: list[str], lineno: int) -> bool:
    for i in range(lineno - 2, max(lineno - 60, -1), -1):
        l = lines[i].strip()
        if "#[test]" in l or "#[cfg(test)]" in l:
            return True
        if l.startswith("mod tests") or l.startswith("mod test"):
            return True
    return False


def check(path: Path, lines: list[str]) -> Generator[Issue, None, None]:
    if _is_test_file(path):
        return

    for lineno, raw in enumerate(lines, start=1):
        stripped = raw.lstrip()
        if stripped.startswith("//"):
            continue
        if _CONST_DEF.match(raw):
            continue
        if _is_test_context(lines, lineno):
            continue

        # ── match "foo" => ──────────────────────────────────────────────────
        for m in _MATCH_STR_ARM.finditer(raw):
            val = m.group(1)
            if _SKIP_VALUES.search(val):
                continue
            yield Issue(
                file=path, line=lineno, col=m.start() + 1,
                severity=Severity.ERROR,
                rule=f"{_RULE_BASE}/no-string-match",
                message=(
                    f'stringly-typed match "{val}" — type-safety paradigm (rust/types): '
                    f"define an enum variant instead of a raw string literal in match arms"
                ),
            )

        # ── == "identifier" ─────────────────────────────────────────────────
        for m in _EQ_STR.finditer(raw):
            val = m.group(1)
            if _SKIP_VALUES.search(val):
                continue
            yield Issue(
                file=path, line=lineno, col=m.start() + 1,
                severity=Severity.ERROR,
                rule=f"{_RULE_BASE}/no-string-compare",
                message=(
                    f'stringly-typed comparison == "{val}" — type-safety paradigm (rust/types): '
                    f"discriminators must be enums or named consts, not raw string literals"
                ),
            )

File: crates/scanner/src/checks/string_states.rs
use std::path::Path;

use regex::Regex;
use std::sync::LazyLock;

use crate::config::Config;
use crate::context::{self, FileContext};
use crate::issue::{Issue, Severity};

static MATCH_ARM_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#""([^"]{2,}?)"\s*=>"#).unwrap()
});
static EQUALITY_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"[!=]=\s*"([^"]{2,}?)""#).unwrap()
});
/// Strings that look like messages/paths rather than identifiers.
static SKIP_VALUE_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(?:\s|[/\\.]|[A-Z]{2}|[?!,])").unwrap()
});

/// Check for stringly-typed match arms and comparisons.
pub fn check(
    file_ctx: &FileContext,
    lines: &[&str],
    _cfg: &Config,
    issues: &mut Vec<Issue>,
    path: &Path,
) {
    if file_ctx.is_test_file {
        return;
    }

    for (line_num, line) in lines.iter().enumerate() {
        let trimmed = line.trim();
        if trimmed.starts_with("//") {
            continue;
        }
        if context::is_const_def(line) {
            continue;
        }
        if context::is_test_context(lines, line_num) {
            continue;
        }

        for cap in MATCH_ARM_RE.captures_iter(line) {
            let val = &cap[1];
            if SKIP_VALUE_RE.is_match(val) {
                continue;
            }
            issues.push(Issue::new(
                path, line_num + 1, 1, Severity::Error,
                "rust/types/no-string-match",
                format!("stringly-typed match \"{val}\" — define an enum variant instead"),
            ));
        }

        for cap in EQUALITY_RE.captures_iter(line) {
            let val = &cap[1];
            if SKIP_VALUE_RE.is_match(val) {
                continue;
            }
            issues.push(Issue::new(
                path, line_num + 1, 1, Severity::Error,
                "rust/types/no-string-compare",
                format!("stringly-typed comparison \"{val}\" — discriminators must be enums or named consts"),
            ));
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::context::Language;

    fn prod_ctx() -> FileContext {
        FileContext { language: Language::Rust, is_test_file: false, is_mother_file: false, is_definition_file: false }
    }

    #[test]
    fn catches_match_arm() {
        let mut issues = Vec::new();
        check(&prod_ctx(), &[r#"    "active" => do_thing(),"#], &Config::default(), &mut issues, Path::new("a.rs"));
        assert_eq!(issues.len(), 1);
        assert!(issues[0].rule_id.contains("string-match"));
    }

    #[test]
    fn catches_equality() {
        let mut issues = Vec::new();
        check(&prod_ctx(), &[r#"if state == "running" {"#], &Config::default(), &mut issues, Path::new("a.rs"));
        assert_eq!(issues.len(), 1);
    }

    #[test]
    fn skips_message_strings() {
        let mut issues = Vec::new();
        check(&prod_ctx(), &[r#"    "file not found" => bail!(),"#], &Config::default(), &mut issues, Path::new("a.rs"));
        assert!(issues.is_empty());
    }

    #[test]
    fn skips_path_strings() {
        let mut issues = Vec::new();
        check(&prod_ctx(), &[r#"if ext == "src/main.rs" {"#], &Config::default(), &mut issues, Path::new("a.rs"));
        assert!(issues.is_empty());
    }
}
